### PR TITLE
20/spectrum bar

### PR DESCRIPTION
**Pull request by hagoptaminian**
fairly big update (made in a series of commits) addressing most outstanding issues. if this works well, app is essentially complete, although additional features, like the alternative view of the electric field (#14), may still be introduced.

one important feature to verify is whether the frequency stated in the app is producing the wavelengths expected in the table below (as explained in #26). (this has been set up to work correctly, and is working in local testing, but other devices/browsers may produce unexpected behavior)

Screen Shot 2020-12-14 at 1 48 25 PM

### CODE REVIEW

**Review by pyphilia — COMMENTED** [review]
The app looks really nice now! Some comments while playing around with it:
- Do we plan to keep both the switch and play/pause button? Maybe the play/pause button only are enough.
- The measuring arrow is a bit shifted corresponding to your grid. Additionally, the grid is redrawn and has a different size depending on whether the sidemenu is open or close.
- Maybe we should ask Phil and Thierry, but why not move the measuring arrow really close to the parallel line (direction -1,0) to ease the measure, and apply the corresponding color described by the spectrum bar depending on how long is the arrow?

**Comment by hagoptaminian** [comment]
thanks Kim :)

- I think we still need both buttons! keep in mind... hitting pause **freezes** the animation (with whatever curves that are there due to the oscillation). whereas **stopping** the oscillation would result in the straight radial lines that are the default when the app starts. 
- two things: (1) this arrow is draggable. so the idea is that the user can drag it to measure a wavelength. (2) yes... the grid is meant to be responsive. thought that would look better given different screen sizes
- let's discuss... specifically, with _"and apply the corresponding color described by the spectrum bar depending on how long is the arrow"_ --> you mean apply the color to the arrow?

File: src/reducers/layout.js [hunk 1]
```diff
@@ -1,16 +1,33 @@
 import {
-  DEFAULT_AMPLITUDE_VALUE,
+  DEFAULT_AMPLITUDE,
```
Comment (juancarlosfarah) [inline src/reducers/layout.js:2]: At some point maybe we should break out the non-layout part of this reducer into a name that makes more sense... `lab`?

File: src/components/common/SideMenu.js [hunk 2]
```diff
@@ -100,11 +138,56 @@ class SideMenu extends React.Component {
         >
           {this.renderDrawerHeader()}
           <div className={classes.contentWrapper}>
-            {this.renderDescription()}
             <LineSelector />
-            <OscillateSwitch />
-            <AmplitudeSlider disabled={!shouldOscillate} />
-            <FrequencyAdjuster />
+            <div className={classes.switchContainer}>
```
Comment (juancarlosfarah) [inline src/components/common/SideMenu.js:142]: I really like the factoring out of the switches and sliders. 👍

**Review by juancarlosfarah — CHANGES_REQUESTED** [review]
Looks good! Just some minor comments in the code and a comment here.

image

Looks like the order in which things are draw causes some lines to be on top of others.

File: src/reducers/layout.js (hunk 1 above)
Comment (hagoptaminian) [reply src/reducers/layout.js:2]: created an issue with this https://github.com/graasp/graasp-app-radiating-charge/issues/28

File: src/components/common/SideMenu.js (hunk 2 above)
Comment (hagoptaminian) [reply src/components/common/SideMenu.js:142]: 👍

**Comment by hagoptaminian** [comment]
> Looks good! Just some minor comments in the code and a comment here.
> 
> image
> 
> Looks like the order in which things are draw causes some lines to be on top of others.

aha you're correct. good call. I noticed this in another component but after looking at the app so much I had stopped noticing that this doesn't look great. looks much nicer after moving!

**Comment by pyphilia** [comment]
@hagoptaminian 

- I understand why you would like to keep the switch. However you can obtain these straight lines as well when you are reseting the simulation. At best, at the reset should put everything to "zero" so that the inputs actually gives straight lines instead of forcing the app to be idle (yes the inputs are disabled but I think it would be more meaningful if the inputs match the simulation when idle).
- Oh didn't notice it was draggable, it's very nice! Maybe an icon beside it (you know this move icon with a cross made of arrows?). But I don't understand the grid, because then it doesn't match the arrow length the same way when the side menu is opened and closed. Wasn't it supposed to give a measuring input?
- If I understand this physic phenomenon correctly, depending on the wavelength, the lines emit different "energy" and thus color corresponding to the spectrum bar. What do you think?

**Review by juancarlosfarah — APPROVED** [review]
🔥

**Comment by hagoptaminian** [comment]
@pyphilia 

- let me think about this 
- it's true the grid changes size, but the arrow and the wavelengths are the same regardless of screen size, so the measurement is still correct. of course, we could fix the gird size as well, and that way the measuring arrow is always the same relative to the grid, regardless of screen size.
- I think the spectrum bar is just a visual aid, so that once the students measure a wavelength, they can see where in the electromagnetic spectrum that wavelength lies. so at higher frequencies you get smaller (UV) wavelengths, which are not visible to the human eye. but I might be misunderstanding your suggestion - let's discuss in person/on zoom

**Pull request merged**